### PR TITLE
Non-matching captures return nil instead of empty string 

### PR DESCRIPTION
**Pull request by will**
This is a breaking change, but brings the API in line with similar regular expression APIs.

Before:

    /f(o)(x?)/.match("the fort") #=> #<MatchData "fo" 1:"o" 2:"">
    /f(o)(x)?/.match("the fort") #=> #<MatchData "fo" 1:"o" 2:"">

After:

    /f(o)(x?)/.match("the fort") #=> #<MatchData "fo" 1:"o" 2:"">
    /f(o)(x)?/.match("the fort") #=> #<MatchData "fo" 1:"o" 2:nil>

### CODE REVIEW

**Comment by ysbaddaden** [comment]
But then... all match data values are nilable, and we must deal it :-(

File: spec/std/regexp_spec.cr [hunk 1]
```diff
@@ -49,11 +49,18 @@ describe "Regex" do
...
+    it "captures empty group" do
+      ("foo" =~ /(?<g1>z?)foo/).should eq(0)
+      $~[1].should eq("")
       $~["g1"].should eq("")
     end
 
+    it "raises exception on optional empty group" do
+      ("foo" =~ /(?<g1>z)?foo/).should eq(0)
+      expect_raises(Exception) { $~[1] }
+      expect_raises(Exception) { $~["g1"] }
```
Comment (jhass) [inline spec/std/regexp_spec.cr]: Does it really raise `Exception`? nothing more specific?

**Comment by asterite** [comment]
@ysbaddaden I initially thought the same! But it seems @will retained the semantic of `[]` raises on nil and `[]?` returns nil if missing.

The only drawback is that now `$1`, `$2`, etc. might raise because they could be nil. These are rewritten to `$~.not_nil![1]`, etc. But maybe we could add `$1?`, `$2?` just for completeness, in cases where groups might be optional.

On the other hand, having missing groups be represented as empty strings isn't that bad too: you can safely assume you will have strings for all groups, you just check with `empty?` if they were matched or not. So I'm not sure what's the best option here.

Others' thoughts?

File: spec/std/regexp_spec.cr (hunk 1 above)
Comment (will) [inline spec/std/regexp_spec.cr]: Looks like it http://carc.in/#/r/2x0

**Comment by will** [comment]
If you have optional things inside a group `(\d?)` and nothing matches, the group still succeeded, so the response is `""`. If the entire group is optional `(\d)?`, and nothing matches, the group failed, so response is `nil`.

I'm doing this regex work because I originally wanted to add username/password support to URI, and having to do `path = $5.empty? ? nil : $5` for every capture group seemed tedious. 

I don't feel super strongly either way. Instead String could have something like sql NULLIF, or failing that a helper function in that class. The part of the patch that switches to `pcre_get_stringnumber` is important though, so there is only one path through named vs numbered groups, as it's easy to add behavior to only one and forget the other.

**Comment by kirbyfan64** [comment]
FYI, Nim very recently did something similar and started returning nil instead of an empty string for a failed match.

**Comment by ysbaddaden** [comment]
I think the behavior @will explained is better. I'm a bit puzzled on the magic variables, thought. Especially the `$1?` notation that would be too much —and introduce more magic vars, even if these are local.

**Comment by asterite** [comment]
We finally discusses this with @waj and we agreed that it's OK to have `nil` for non-matching captures. To deal with this we'll also add `$1?`, `$2?`, etc. Although these might look ugly, remember that these are just syntax sugar. `$1` is just `$~.not_nil![1]` while `$1?` is just `$~.not_nil![1]?`.

**Pull request merged**